File: handlers/scoresa.go
package handlers

import (
	"encoding/json"
	"github.com/adrianfish/dialang-web/datacapture"
	"github.com/adrianfish/dialang-web/models"
	"github.com/adrianfish/dialang-web/scoring"
	"github.com/adrianfish/dialang-web/session"
	"log"
	"net/http"
	"strings"
)

func ScoreSA(w http.ResponseWriter, r *http.Request) {

	dialangSession := session.SessionManager.Get(r.Context(), "session").(models.DialangSession)

	if dialangSession.TES.TL == "" || dialangSession.TES.Skill == "" {
		w.WriteHeader(http.StatusInternalServerError)
		return
	}

	decoder := json.NewDecoder(r.Body)
	var saResponses map[string]string
	if err := decoder.Decode(&saResponses); err != nil {
		w.WriteHeader(http.StatusInternalServerError)
		return
	}

	skill := dialangSession.TES.Skill

	responses := map[string]bool{}
	for name, v := range saResponses {
		if strings.HasPrefix(name, "statement:") {
			wid := strings.Split(name, ":")[1]
			responses[wid] = v == "yes"
		}
	}

	saPPE, saLevel, err := scoring.GetSaPPEAndLevel(skill, responses)
	if err != nil {
		log.Printf("Failed to score self assessment for skill %s\n", skill)
		w.WriteHeader(http.StatusInternalServerError)
		return
	}

	dialangSession.SaPPE = saPPE
	dialangSession.SaSubmitted = true
	dialangSession.SaLevel = saLevel

	session.SessionManager.Put(r.Context(), "session", dialangSession)

	datacapture.LogSAResponses(&dialangSession, responses)
	datacapture.LogSAScores(&dialangSession)

	json.NewEncoder(w).Encode(map[string]any{"saPPE": saPPE, "saLevel": saLevel})
	return
}

File: src/routes/scoresa.ts
import { Context }  from "@hono";
import { getSessionId } from "../utils/utils.ts";
import { getSaPPEAndLevel } from "../scoring/scoring.ts";
import type { DialangSession } from "../types.ts";
import type { Storage } from "../storage/storage.ts";

export async function scoreSA(c: Context, storage: Storage): Promise<Response> {

  const sessionId = getSessionId(c);
  if (!sessionId) {
    console.error("Failed to get session id");
		c.status(500);
		return c.html("");
  }

  const session: DialangSession | null = await storage.getSession(sessionId);
  if (!session) {
		console.error(`No session for id ${sessionId}. Returning 500 ...`)
		c.status(500);
		return c.html("");
  }

	if (!session.tl || !session.skill) {
		c.status(500);
		return c.html("");
	}

  const body = await c.req.json();

  const responses: Record<string, boolean> = Object.fromEntries(
      Object.entries(body)
        .filter(([k, v]) => k.startsWith("statement:"))
        .map(([k, v]) => [ k.split(":")[1], v === "yes"])
    );

  const [ppe, level, err ] = await getSaPPEAndLevel(session.skill, responses, storage);
	if (err) {
		console.error(`Failed to score self assessment for skill ${session.skill}`);
    c.status(500);
		return c.html("");
	}

	session.saPPE = ppe as number;
	session.saSubmitted = true;
	session.saLevel = level as string;

  storage.saveSession(sessionId, session);

  /*
	datacapture.LogSAResponses(&dialangSession, responses)
	datacapture.LogSAScores(&dialangSession)
  */

	storage.logSaScores(session);

	return c.json({ ppe, level });
}
